addon/-private/system/model/internal-model.js

```diff
@@ -490,11 +492,24 @@ export default class InternalModel {
   unloadRecord() {
     this.send('unloadRecord');
     this.dematerializeRecord();
-    Ember.run.schedule('destroy', this, '_checkForOrphanedInternalModels');
+    if (this._scheduledDestroy === null) {
+      this._scheduledDestroy = run.schedule('destroy', this, '_checkForOrphanedInternalModels');
+    }
+  }
+
+  cancelDestroy() {
+    assert(`You cannot cancel the destruction of an InternalModel once it has already been destroyed`, !this.isDestroyed);
+
+    if (this._isDematerializing === true) {
+      this._isDematerializing = false;
+      run.cancel(this._scheduledDestroy);
+      this._scheduledDestroy = null;
```

Comment:
Good catch. The intent was to avoid unnecessary work if we weren't in a state that needed to be cleaned up, but this isn't the correct check for that. Moved to just always resetting all three things.